src/http/url.cc

```diff
@@ -91,6 +91,27 @@ bool url_decode(const std::string_view& in, sstring& out) {
     return true;
 }
 
+bool path_decode(const std::string_view& in, sstring& out) {
+    size_t pos = 0;
+    sstring buff(in.length(), 0);
+    for (size_t i = 0; i < in.length(); ++i) {
+        if (in[i] == '%') {
+            if (i + 3 <= in.size()) {
+                buff[pos++] = hexstr_to_char(in, i + 1);
+                i += 2;
+            } else {
+                return false;
+            }
+        } else {
+            buff[pos++] = in[i];
+        }
+    }
+    buff.resize(pos);
+    out = buff;
+    return true;
```

Comment:
I think it's worth sharing some of this implementation with `url_decode`.

Reply:
I created a templated `decode` method now. It's templated on whether to replace plus or not. Let me know if you had something else in mind. I was also considering extracting just the percent decoding instead of templating, but that's hard to extract because it updates so many variables.